File: gascapi.py
# // TODO: This is mostly just the quickstart....
# // This needs to be reafactored with the functionality in ./extra/mysql2mongo.py
# // https://github.com/google/google-api-python-client/

from __future__ import print_function
import os, sys
import httplib2
from apiclient import discovery
from apiclient import errors
import oauth2client
from oauth2client import client
from oauth2client import tools

try:
    import argparse
    flags = argparse.ArgumentParser(parents=[tools.argparser]).parse_args()
    flags.noauth_local_webserver = True
except ImportError:
    flags = None

import json
with open('package.json') as data:
    pkg = json.load(data)

SCRIPT_ID = CLIENT_SECRET = APPLICATION_NAME = CREDENTIAL_PATH = None

APPLICATION_NAME = "{0} ({1}-v{2})".format(pkg['description'], pkg['name'], pkg['version'])

path = pkg['config']['config_dir']
dirs = os.listdir(path)
for file in dirs:
    file = os.path.join(path, file)
    if file.endswith('script.id'):
        with open(file, 'r') as data:
            SCRIPT_ID = data.read()
    if file.endswith('auth_secret'):
        CLIENT_SECRET = file

# base_path = os.path.expanduser('~')   # original user home dir
base_path = os.path.dirname(os.path.realpath(__file__))  # modified project home dir
result = os.path.join(base_path, pkg['config']['credential_dir'])
if not os.path.exists(result):
    os.makedirs(result)
CREDENTIAL_PATH = os.path.join(result, "{0}.json".format(__file__.replace('.', '-')))
SCOPES = ' '.join(pkg['config']['scopes'])
# print (SCRIPT_ID, CLIENT_SECRET, APPLICATION_NAME, CREDENTIAL_PATH, SCOPES)
# import sys
# sys.exit()


def get_credentials():
    """Gets valid user credentials from storage.

    If nothing has been stored, or if the stored credentials are invalid,
    the OAuth2 flow is completed to obtain the new credentials.

    Returns:
        Credentials, the obtained credential.
    """
    store = oauth2client.file.Storage(CREDENTIAL_PATH)
    credentials = store.get()
    if not credentials or credentials.invalid:
        flow = client.flow_from_clientsecrets(CLIENT_SECRET, SCOPES)
        flow.user_agent = APPLICATION_NAME
        if flags:
            credentials = tools.run_flow(flow, store, flags)
        else:  # Needed only for compatibility with Python 2.6
            credentials = tools.run(flow, store)
        print('Storing credentials to ' + CREDENTIAL_PATH)
    return credentials


def display_error(response):
    """Displays the error results and stacktrace for an unsuccessful request

    Displays the error results and stacktrace for an unsuccessful request
    """
    # The API executed, but the script returned an error.

    # Extract the first (and only) set of error details. The values of
    # this object are the script's 'errorMessage' and 'errorType', and
    # an list of stack trace elements.
    error = response['error']['details'][0]
    print("Script error message: {0}".format(error['errorMessage']))

    if 'scriptStackTraceElements' in error:
        # There may not be a stacktrace if the script didn't start
        # executing.
        print("Script error stacktrace:")
        for trace in error['scriptStackTraceElements']:
            print("\t{0}: {1}".format(trace['function'],
                trace['lineNumber']))


def get_service():
    credentials = get_credentials()
    http = credentials.authorize(httplib2.Http())
    service = discovery.build('script', 'v1', http=http)
    return service


def encode_result(response):
    results = response['response'].get('result', {})
    return json.dumps(results)


def execute(request, callback):
    if ((not CLIENT_SECRET) or (not os.path.isfile(CLIENT_SECRET)) or (os.path.getsize(CLIENT_SECRET) < 23)):
        print("auth_secret must exist and must not be empty")
        sys.exit(404)

    if ((not SCRIPT_ID) or len(SCRIPT_ID) < 42):
        print("script.id must exist and must not be empty ")
        sys.exit(404)

    service = get_service()
    try:
        response = service.scripts().run(body=request, scriptId=SCRIPT_ID).execute()

        if 'error' in response:
            display_error(response)
        else:
            return callback(response)

    except errors.HttpError as e:
        # The API encountered a problem before the script started executing.
        print(e.content)


# --------------------
def addContactToGroup(person):
    request = {"function": "addContactToGroup", "parameters": [person]}
    return execute(request, encode_result)


def addContactToGroupByName(first_name, last_name, email, group_name):
    request = {"function": "addContactToGroup", "parameters":
        [{"f": first_name, "l": last_name, "e": email, "g": group_name}]}
    return execute(request, encode_result)


def getContactList(group_name):
    request = {"function": "getContactList", "parameters": [{"g": group_name}]}
    return execute(request, encode_result)


def addGroup(group_name):
    request = {"function": "addGroup", "parameters": [{"g": group_name}]}
    return execute(request, encode_result)


def main():
    # default function to verify configuration
    print(getContactList('2011 South Calgary Garden'))


if __name__ == '__main__':
    main()

File: gascapi.js

// TODO: This is mostly just the quickstart.... 
// not sure how the node/js flavor of this interface will be used

var fs = require('fs');
var readline = require('readline');
var google = require('googleapis');
var googleAuth = require('google-auth-library');

var CLIENT_SECRET, SCRIPT_ID, APPLICATION_NAME, CREDENTIAL_PATH, CREDENTIAL_DIR, SCOPES;
CLIENT_SECRET = SCRIPT_ID = APPLICATION_NAME = CREDENTIAL_PATH = CREDENTIAL_DIR = SCOPES = undefined;

function loadConfiguration(callback) {
  fs.readFile('package.json', function(err, content) {
    if (err) {
      console.log('Error loading package.json: ' + err);
      return;
    }

    var pkg = JSON.parse(content);
    var path = pkg.config.config_dir;

    APPLICATION_NAME = pkg.description.concat(' (', pkg.name, '-v',  pkg.version, ')');
    CREDENTIAL_DIR = pkg.config.credential_dir;
    CREDENTIAL_PATH = CREDENTIAL_DIR.concat('/', module.filename.split('/').pop().replace('.', '-'), '.json');
    SCOPES = pkg.config.scopes;

    fs.readdir( path, function( err, files ) {
      if( err ) {
        console.error( "Could not list the directory.", err );
        return;
      } 

      files.forEach( function( file ) {
        file = path.concat('/', file);
        if (file.endsWith('script.id')) {
          SCRIPT_ID = fs.readFileSync(file).toString('UTF-8');
        }
        if (file.endsWith('auth_secret')) {
          CLIENT_SECRET = file;
        }
      });

      // console.log(SCRIPT_ID, CLIENT_SECRET, APPLICATION_NAME, CREDENTIAL_DIR, CREDENTIAL_PATH, SCOPES);
      // process.exit();
      callback();
    });
  });
}

/**
 * Store token to disk be used in later program executions.
 *
 * @param {Object} token The token to store to disk.
 */
function storeToken(token) {
  try {
    fs.mkdirSync(CREDENTIAL_DIR);
  } catch (err) {
    if (err.code !== 'EEXIST') {
      throw err;
    }
  }
  fs.writeFile(CREDENTIAL_PATH, JSON.stringify(token));
  console.log('Token stored to ' + CREDENTIAL_PATH);
}

/**
 * Get and store new token after prompting for user authorization, and then
 * execute the given callback with the authorized OAuth2 client.
 *
 * @param {google.auth.OAuth2} oauth2Client The OAuth2 client to get token for.
 * @param {getEventsCallback} callback The callback to call with the authorized
 *     client.
 */
function getNewToken(oauth2Client, callback) {
  var authUrl = oauth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: SCOPES
  });
  console.log('Authorize this app by visiting this url: ', authUrl);
  var rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  rl.question('Enter the code from that page here: ', function(code) {
    rl.close();
    oauth2Client.getToken(code, function(err, token) {
      if (err) {
        console.log('Error while trying to retrieve access token', err);
        return;
      }
      oauth2Client.credentials = token;
      storeToken(token);
      callback(oauth2Client);
    });
  });
}

/**
 * Create an OAuth2 client with the given credentials, and then execute the
 * given callback function.
 *
 * @param {Object} credentials The authorization client credentials.
 * @param {function} callback The callback to call with the authorized client.
 */
function authorize(credentials, callback) {
  var clientSecret = credentials.installed.client_secret;
  var clientId = credentials.installed.client_id;
  var redirectUrl = credentials.installed.redirect_uris[0];
  var auth = new googleAuth();
  var oauth2Client = new auth.OAuth2(clientId, clientSecret, redirectUrl);

  // Check if we have previously stored a token.
  fs.readFile(CREDENTIAL_PATH, function(err, token) {
    if (err) {
      getNewToken(oauth2Client, callback);
    } else {
      oauth2Client.credentials = JSON.parse(token);
      callback(oauth2Client);
    }
  });
}

function execMethod(auth, name, params, callback) {
  var script = google.script('v1');
  var request = {};
  request.function = name;
  if(params) {
    // TODO: Ensure params is an array...
    request.parameters = params;
  }

  // Make the API request. The request object is included here as 'resource'.
  script.scripts.run({
    auth: auth,
    resource: request,
    scriptId: SCRIPT_ID
  }, function(err, resp) {
    if (err) {
      // The API encountered a problem before the script started executing.
      console.log('The API returned an error: ' + JSON.stringify(err));
      return;
    }
    if (resp.error) {
      // The API executed, but the script returned an error.
      // Extract the first (and only) set of error details. The values of this
      // object are the script's 'errorMessage' and 'errorType', and an array
      // of stack trace elements.
      var error = resp.error.details[0];
      console.log('Script error message: ' + error.errorMessage);
      console.log('Script error stacktrace:');

      if (error.scriptStackTraceElements) {
        // There may not be a stacktrace if the script didn't start executing.
        for (var i = 0; i < error.scriptStackTraceElements.length; i++) {
          var trace = error.scriptStackTraceElements[i];
          console.log('\t%s: %s', trace.function, trace.lineNumber);
        }
      }
    } else {
      callback(resp);
    }

  });
}

function onConfigLoaded() {
  // Load client secrets from a local file.
  fs.readFile(CLIENT_SECRET, function(err, content) {
    if (err) {
      console.log('Error loading client secret file: ' + err);
      return;
    }
    // Authorize a client with the loaded credentials, then call the Google Apps Script Execution API.
    authorize(JSON.parse(content), function(auth) {

      execMethod(auth,
        'getContactList', 
        { g: '2011 South Calgary Garden' }, 
        function(resp) {
          var result = JSON.stringify(resp.response.result);
          console.log(result);
        });

    });
  }); 
}

String.prototype.endsWith = function(suffix) {
    return this.indexOf(suffix, this.length - suffix.length) !== -1;
};

// ******************
loadConfiguration(onConfigLoaded);